src/conduits/RwaInputConduit3.t.sol

```diff
@@ -0,0 +1,482 @@
+    }
+
+    function testPushAmountWhenHaveSomeDaiBalanceGetExactAmount() public {
+        dai.mint(address(inputConduit), 100 ether);
+        assertEq(dai.balanceOf(address(inputConduit)), 100 ether);
+
+        assertEq(usdx.balanceOf(me), USDX_MINT_AMOUNT);
+        assertEq(usdx.balanceOf(address(inputConduit)), 0);
+        assertEq(usdx.balanceOf(address(joinA)), 0);
+
+        usdx.transfer(address(inputConduit), 500 * USDX_BASE_UNIT);
+
+        assertEq(usdx.balanceOf(me), USDX_MINT_AMOUNT - 500 * USDX_BASE_UNIT);
+        assertEq(usdx.balanceOf(address(inputConduit)), 500 * USDX_BASE_UNIT);
+
+        assertEq(testUrn.balance(address(dai)), 0);
+
+        vm.expectEmit(true, true, false, false);
+        emit Push(address(testUrn), 500 ether);
+        inputConduit.push(500 * USDX_BASE_UNIT);
+
+        assertEq(usdx.balanceOf(address(joinA)), 500 * USDX_BASE_UNIT);
+        assertEq(usdx.balanceOf(address(inputConduit)), 0);
+        assertEq(testUrn.balance(address(dai)), gemToDai(500 * USDX_BASE_UNIT));
+        assertEq(dai.balanceOf(address(inputConduit)), 100 ether);
+    }
+
+    function testPushAmountWhenHaveSomeDaiBalanceGetAll() public {
+        dai.mint(address(inputConduit), 100 ether);
+        assertEq(dai.balanceOf(address(inputConduit)), 100 ether);
+
+        assertEq(usdx.balanceOf(me), USDX_MINT_AMOUNT);
+        assertEq(usdx.balanceOf(address(inputConduit)), 0);
+        assertEq(usdx.balanceOf(address(joinA)), 0);
+
+        usdx.transfer(address(inputConduit), 500 * USDX_BASE_UNIT);
+
+        assertEq(usdx.balanceOf(me), USDX_MINT_AMOUNT - 500 * USDX_BASE_UNIT);
+        assertEq(usdx.balanceOf(address(inputConduit)), 500 * USDX_BASE_UNIT);
+
+        assertEq(testUrn.balance(address(dai)), 0);
+
+        vm.expectEmit(true, true, false, false);
+        emit Push(address(testUrn), 500 ether);
+        inputConduit.push();
+
+        assertEq(usdx.balanceOf(address(joinA)), 500 * USDX_BASE_UNIT);
+        assertEq(usdx.balanceOf(address(inputConduit)), 0);
+        assertEq(testUrn.balance(address(dai)), gemToDai(500 * USDX_BASE_UNIT));
+        assertEq(dai.balanceOf(address(inputConduit)), 100 ether);
+    }
+
+    function testPushAmountFuzz(uint256 amt) public {
+        uint256 usdxBalanceBefore = usdx.balanceOf(me);
+        usdx.transfer(address(inputConduit), usdxBalanceBefore);
+        assertEq(usdx.balanceOf(me), 0);
+
+        uint256 usdxCBalanceBefore = usdx.balanceOf(address(inputConduit));
+        uint256 urnUsdxBalanceBefore = usdx.balanceOf(address(testUrn));
+
+        amt = bound(amt, 1 * USDX_BASE_UNIT, usdxCBalanceBefore);
+
+        vm.expectEmit(true, true, false, false);
+        emit Push(address(testUrn), amt * USDX_DAI_DIF_DECIMALS);
+        inputConduit.push(amt);
+
+        assertEq(usdx.balanceOf(address(inputConduit)), usdxCBalanceBefore - amt);
+        assertEq(testUrn.balance(address(dai)), urnUsdxBalanceBefore + gemToDai(amt));
```

Comment:
There is something wrong here, it's adding up a USDX balance with a DAI one.

Reply:
The only way this is passing is because `urnUsdxBalanceBefore` is zero, so maybe to make the fuzzing more interesting we should also get the DAI one and make it random as well.
Also why do we have `TestUrn` as a contract only for returning a getter value that doesn't need to be wrapped at all.
In fact the usdx balance is being got directly meanwhile the DAI one goes through the contract function.

Reply:
Btw using `forge` we do not even need wrapped contract for writing actions either as we can impersonate (as it has been used in other parts).